internal/inspect/internal/policy/inspect.go

```diff
@@ -0,0 +1,324 @@
+// Copyright 2021-2024 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package policy
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/google/cel-go/common/ast"
+
+	policyv1 "github.com/cerbos/cerbos/api/genpb/cerbos/policy/v1"
+	responsev1 "github.com/cerbos/cerbos/api/genpb/cerbos/response/v1"
+	runtimev1 "github.com/cerbos/cerbos/api/genpb/cerbos/runtime/v1"
+	"github.com/cerbos/cerbos/internal/compile"
+	"github.com/cerbos/cerbos/internal/inspect/internal"
+	"github.com/cerbos/cerbos/internal/namer"
+	"github.com/cerbos/cerbos/internal/policy"
+)
+
+func New() *Inspect {
+	return &Inspect{
+		inspections: make(map[string]*internal.Inspection),
+	}
+}
+
+type Inspect struct {
+	inspections map[string]*internal.Inspection
+}
+
+func (i *Inspect) Inspect(p *policyv1.Policy) error {
+	if p == nil {
+		return fmt.Errorf("policy is nil")
+	}
+
+	id := namer.PolicyKey(p)
+	conditions, err := i.inspectConditions(p)
+	if err != nil {
+		return fmt.Errorf("failed to inspect conditions of the policy %s: %w", id, err)
+	}
+
+	i.inspections[id] = &internal.Inspection{
+		Actions:    policy.ListActions(p),
+		Conditions: conditions,
+		Imports:    i.inspectImports(p),
+		Variables:  policy.ListVariables(p),
+	}
+
+	return nil
+}
+
+func (i *Inspect) Results() (map[string]*responsev1.InspectPoliciesResponse_Result, error) {
```

Comment:
It took me a long time to see what was going on here because of all the deeply nested loops and if conditions 🙂 I think we can we try to split this into smaller, more understandable bits.

The `Inspect` method above should take a policy and do the following:
- If it's an `ExportVariables` policy, all its variables are resolved already. Just add the list of sorted variable names to `variables` as `KIND_EXPORT`.
- For any other policy type:
  - Extract the list of variables referenced by its conditions
  - Add the local variables to `variables` list with kind `KIND_LOCAL` and, if any of them are referenced by the conditions, mark as "used". (I just realised we need another flag in the protobuf to mark whether a variable is used or not.)
  - For the remainder of referenced variables that are not local variables, add them to a temporary field `toResolve`

In the `Results` method: 
  - Go through the list of policies that have a non-empty `toResolve` field 
     - If they import any `ExportVariables` load them (if we haven't already loaded them during our traversal) and resolve the `KIND_IMPORT` variables.
     - Any variable that's not defined in either local or imported variables should be marked as `KIND_UNDEFINED` and have `null` as the value.

Reply:
Since we do not have the option to show unused variables in the compiled policy stores, I opted to only list the used variables in the `inspect.Policy` too. If we introduce a `used` field, we cannot provide this information in the compiled policy stores which could be confusing.

> Any variable that's not defined in either local or imported variables should be marked as KIND_UNDEFINED and have null as the value.

I think we should always assume the compilation step is successful, and this step is not possible in that case, no?

I have put some effort to modify the code to follow the steps you've written and added some more steps in between to cover what the current implementation does. All in all, I think it did not change much in terms of readability though I have not completed it.

The actual complexity comes from resolving the variables used in the conditions. For example the following is not as simple because I also remove the unused variables from the export variables.

> If it's an ExportVariables policy, all its variables are resolved already. Just add the list of sorted variable names to variables as KIND_EXPORT.

Reply:
> Since we do not have the option to show unused variables in the compiled policy stores, I opted to only list the used variables in the `inspect.Policy` too. If we introduce a `used` field, we cannot provide this information in the compiled policy stores which could be confusing.

Everything would be "used" in compiled policies because they are in an optimized format and we don't have the raw source. I don't think that would be confusing for users. The output of `inspect` would be consistent for the store in use because Cerbos can only have one store anyway.   


> I think we should always assume the compilation step is successful, and this step is not possible in that case, no?

If you're talking about `condition.Compile`, that doesn't detect undefined variables because `V` is a dynamic container and it has no idea what it could contain at runtime.


> I also remove the unused variables from the export variables.

If I inspect an `ExportVariables` policy with 10 variables but none of those are used by any other policies, do I see 10 variables or 0? It should be 10 when inspecting that policy explicitly. For all other policy types, if they import any variables, we should list all of them along with whether they are actually used or not.

Reply:
> If I inspect an ExportVariables policy with 10 variables but none of those are used by any other policies, do I see 10 variables or 0?

The result was 0 in this case.